### DISCUSSION TITLE
Long-term statistics should be on a separate database!

### DISCUSSION THREAD

**Discussion by energywave** [home-assistant/architecture#654] [General] [answered]
In my opinion Long-term statistics tables should be on a separate database than the main one.

If you have to delete DB for corruption or other problems **you'll loose all your long term statistics with months of precious data**!

**The database is written and rewritten very frequently so the corruption is not an infrequent event.** (at least when using the default SQLite DB)
If the databases are two, however, the last term statistics DB would get written only once an hour when updating statistics, greatly reducing the corruption probability.

Furthermore, by dividing them, it will be possible to delete the normal DB leaving long term statistics intact.
It would be also possible to backup the long term statistics and restore it in case of disaster recovering with ease.

Obviously when using other DB than default (MariaDB, ...) two connection should be configured. This way you can optimize DB parameters and virtually use even two separate servers.

I believe that splitting the database will be the best option we have for data integrity.

**Comment by frenck** [comment]
How do you propose this is done / how are you going to implement this?

The reason I ask, is that it seems like a feature request rather than an architectural proposal you are going to implement? This is mainly driven by the opening line: "In my opinion"

**Reply by energywave** [reply]
Ok Frenck, someone suggested me to post that here but ok, I'm a developer but not in python (damn!) so I'll not implement it at all. Got the point, I'll delete this and add a Feature Request as you said. Hoping you'll read this comment before I'll delete it.
Sorry to bother.

**Comment by frenck** [comment, marked as answer]
As confirmed above, this is a feature request, not an architectural proposal. Please use the community forums instead.

Thanks 👍

**Reply by energywave** [reply]
Sorry for the confusion. I've opened a Feature Request here: https://community.home-assistant.io/t/long-term-statistics-on-a-separate-database/353280
Thank you for all the clarifications.

**Comment by bdraco** [comment]
Have you seen database corruption happen in a build with these two changes?

https://github.com/home-assistant/core/pull/56499
https://github.com/home-assistant/core/pull/49282

**Reply by emontnemery** [reply]
@energywave To be clear, SQLite database is absolutely not expected on a system with working hardware on recent versions of Home Assistant. Frequent culprits are bad power supplies or bad SD-cards.
Please be sure to open an issue and tag it with "integration-recorder" if you see this problem.

**Reply by energywave** [reply]
I'm on 2021.11.1. No, I never experienced database corruption on its own. But I'm reading that's happening a lot to people mostly for the causes @emontnemery cited. Some other people delete DB to clean and resolve some strange behavior (I didn't investigate furthermore on that).
But the main thing is that if the db get corrupted you loose all long term statistics that you're storing carefully from months and months.
So I think that some kind of action should be done. DB division is the best one for me. The other is to backup only long term tables with the backup (aka snapshot).
All that in order to protect long term data of people that's not thinking a lot about what happens under the hood. Home Assistant is getting easier and easier to use so it attract less technical users over time. Every user that loose his long term data is a user that will be angry. And that's a pity in the aim of growing this beautiful piece of software :)
